### DISCUSSION TITLE
[Bug]: resolver doesn't seem to be keeping in sync with form state

### DISCUSSION THREAD

**Discussion by marr** [react-hook-form/react-hook-form#6348] [Q&A]
### Version Number

7.13.0

### Codesandbox/Expo snack

https://codesandbox.io/s/blissful-framework-d9qjy?file=/src/App.js

### Steps to reproduce

1. Change number value in numerical input.
2. See `isValid` set to _false_.
3. See empty `errors` object.

Note: something else is going on with the autocomplete in this example. Changing the value of the autocomplete is breaking the page. I'm working on fixing that.

### Expected behaviour

Changing the number in the input should return an errors object containing `{ arr: 'foo' }`. I am using a form-level yup validation to achieve this.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
empty errors not equal valid, valid is for an entire form.

**Reply by marr** [reply]
I’m not sure what you mean here. Errors is a form level prop as well as isValid. Shouldn’t those two be in sync?

**Reply by bluebill1049** [reply]
hook form focuses a single input at a time, so we don't return the entire form level errors.

**Reply by bluebill1049** [reply]
eg: 

```
isValid === resolver() result
errors === resolver[errors.name] (depend on which field you are interactive with)
```

if you can reproduce that your entire form is valid, but formState is still inValid, then we have a bug.

**Reply by marr** [reply]
Does that mean there is no way to return a form level error? I read up on using yup to achieve this, and I can confirm it returns an error object containing the path that I specify. However, from what you are saying it seems like since I'm not actually interacting with that field, the error won't be set?
As the code sandbox that I made shows, the yup `test` is returning a custom error with the path set to `"arr"` but since I'm interacting with the `count` field, its not being set on the `errors` object.

**Reply by marr** [reply]
Here is the sandbox simulating the `schema.validate` call that I am trying to make work: https://codesandbox.io/s/hopeful-khorana-9mrti?file=/src/index.js.
There is no `path` set on the error, but as I explained above, when I set the `path`, the form `error` object doesn't seem to reflect the path that is returning the ValidationError. Also, when the `test` function returns true, I still see `error` object populated with the previous error if its from a different field.

**Reply by bluebill1049** [reply]
> Does that mean there is no way to return a form level error?

At the moment not, hook form always return an error with which the input user is interacting with, only return the full form error when the user hit the submit button, however, we can have an option to always return form-level error, however, this will be trigger re-render at each key stroke just like a fully controlled form.

**Reply by marr** [reply]
Thanks for letting me know Bill. I think a setting like that would be helpful, as there might be many people needing to build validation around multiple inputs. As in, Input A can affect Input B and Input B can affect Input A. Changing either input can revalidate/invalidate the entire form, so it may be good to support that.

**Reply by bluebill1049** [reply]
Oh that's 'deps', check the latest release note.

**Reply by marr** [reply]
Thanks Bill, and congrats on the new release. I checked out the new `deps` prop. I'm not sure it would work in my case, where I'm relying on the `yupResolver` prop. I fixed my sandbox here https://codesandbox.io/s/blissful-framework-d9qjy?file=/src/App.js. I don't know of any way to achieve what I need.

**Reply by bluebill1049** [reply]
perhaps we should consider an option to return formLevel, it's terrible for form performance, but maybe give developers the option.

**Comment by wdfinch** [comment]
@marr I made a [feature request](https://github.com/react-hook-form/react-hook-form/issues/4356) for this a while back. While the validation you run via yup, joi, etc. will produce the correct value for `formState.isValid`, `formState.errors` will only contain the errors that align with actual inputs. So array validation will **not** produce an error key.

A workaround here is to write a custom resolver or to add hidden inputs that hold the array errors. So for example you could have a hidden input called `arrayErrors` and then set that to invalid when the array is not what you want.

@bluebill1049 I don't think the `deps` feat would work here since this just affects the order of validation not which keys are added to the `formState.errors` object.

I like the idea of adding in form-level errors. The way I see it working is any yup, joi, etc. errors that don't align with a registered input would just be added to a separate object called `formLevel` errors or something. Not sure why this is bad for performance. The only real performance costs I see are:

1. Whether you want your schema validation to exit early (stop validating when an error is found) which we already can do.
2. Writing expensive validation which is outside the scope of the library and on the user. 

We might want to make the current behavior more clear in the docs as well. I think there have been many bug reports because people are unaware of how this works.

**Reply by bluebill1049** [reply]
thanks for the feedback @wdfinch let's arrange a RFC after the `replace` API.

**Reply by wdfinch** [reply]
Sounds good!
